pkg/server/singleprocess/service_pipeline_test.go

```diff
@@ -1077,6 +1077,332 @@ func TestServicePipeline_Run(t *testing.T) {
+		var allJobs []*pb.Job
+		for _, id := range resp.AllJobIds {
+			require.Contains(resp.JobMap, id)
+			names = append(names, resp.JobMap[id].Step)
+
+			j, err := client.GetJob(ctx, &pb.GetJobRequest{JobId: id})
+			require.NoError(err)
+			allJobs = append(allJobs, j)
+		}
+		require.Equal([]string{"root", "B", "C", "D", "E", "F", "G"}, names)
+
+		// Loop through all Job Ids returned from RunPipeline and verify that the
+		// order lines up with the expected order of the pipeline for the test.
+		for i, job := range allJobs {
+			require.Equal(job.Pipeline.PipelineId, pipeline.Id)
+			require.Equal(job.Pipeline.PipelineName, pipeline.Name)
+
+			switch i {
+			case 0:
+				require.Equal(job.Pipeline.Step, "root")
+			case 1:
+				require.Equal(job.Pipeline.Step, "B")
+			case 2:
+				require.Equal(job.Pipeline.Step, "C")
+			case 3:
+				require.Equal(job.Pipeline.Step, "D")
+			case 4:
+				require.Equal(job.Pipeline.Step, "E")
+			case 5:
+				require.Equal(job.Pipeline.Step, "F")
+			case 6:
+				require.Equal(job.Pipeline.Step, "G")
+			}
+		}
+
+	})
+
+	t.Run("returns job ids in expected order for an embedded pipeline", func(t *testing.T) {
+		require := require.New(t)
+		ctx := context.Background()
+
+		// Create our server
+		impl, err := New(WithDB(testDB(t)))
+		require.NoError(err)
+		client := server.TestServer(t, impl)
+
+		// Initialize our app
+		TestApp(t, client, serverptypes.TestJobNew(t, nil).Application)
+
+		// Create our pipeline
+		pipeline := serverptypes.TestPipeline(t, nil)
+		pipeline.Steps["B"] = &pb.Pipeline_Step{
+			Name:      "B",
+			DependsOn: []string{"root"},
+			Kind: &pb.Pipeline_Step_Exec_{
+				Exec: &pb.Pipeline_Step_Exec{
+					Image: "hashicorp/waypoint",
+				},
+			},
+		}
+		pipeline.Steps["C"] = &pb.Pipeline_Step{
+			Name:      "C",
+			DependsOn: []string{"B"},
+			Kind: &pb.Pipeline_Step_Exec_{
+				Exec: &pb.Pipeline_Step_Exec{
+					Image: "hashicorp/waypoint",
+				},
+			},
+		}
+
+		// Having two embedded pipeline step references is what is causing the bug
+		// at the moment, i.e. https://github.com/hashicorp/waypoint/issues/3869
+		pipeline.Steps["Embed"] = &pb.Pipeline_Step{
+			Name:      "Embed",
+			DependsOn: []string{"C"},
+			Kind: &pb.Pipeline_Step_Pipeline_{
+				Pipeline: &pb.Pipeline_Step_Pipeline{
+					Ref: &pb.Ref_Pipeline{
+						Ref: &pb.Ref_Pipeline_Id{
+							Id: "embed",
+						},
+					},
+				},
+			},
+		}
+		pipeline.Steps["AnotherEmbed"] = &pb.Pipeline_Step{
+			Name:      "AnotherEmbed",
+			DependsOn: []string{"Embed"},
+			Kind: &pb.Pipeline_Step_Pipeline_{
+				Pipeline: &pb.Pipeline_Step_Pipeline{
+					Ref: &pb.Ref_Pipeline{
+						Ref: &pb.Ref_Pipeline_Id{
+							Id: "twoembed",
+						},
+					},
+				},
+			},
+		}
+
+		// Create, should get an ID back
+		pipeResp, err := client.UpsertPipeline(ctx, &pb.UpsertPipelineRequest{
+			Pipeline: pipeline,
+		})
+		require.NoError(err)
+
+		// Create another pipeline that references the first one
```

Comment:
The pipeline created on L1275 looks like it contains two exec steps, and doesn't reference the first pipeline (I assume "the first pipeline" is referring to the one created on L1218).

Reply:
These are just bad copy/paste comments 😄 Will remove

Reply:
Gotcha! I thought that maybe these pipelines needed to refer to the other one, and that this test _could_ have been misleadingly passing for that reason 😄